decisions/0012-type-inference.md

```diff
@@ -0,0 +1,220 @@
+
+## Decisions
+
+### Route exports API
+
+Keep the route module export API as is.
+Route modules should continue to export separate values for `loader`, `clientLoader`, `action`, `ErrorBoundary`, `default` component, etc.
+That way standard transforms like treeshaking and React Fast Refresh (HMR) work out-of-the-box.
+
+Additionally, this approach introduces no breaking changes allowing Remix users to upgrade to React Router v7 more easily.
+
+### Pass path params, loader data, and action data as props
+
+Hooks like `useParams`, `useLoaderData`, and `useActionData` are defined once in `react-router` and are meant to be used in _any_ route.
+Without any coupling to a specific route, inferring route-specific types becomes impossible and would necessitate user-supplied generics.
+
+Instead, each route export should be provided route-specific args:
+
+```ts
+// Imagine that we *somehow* had route-specific types for:
+// - LoaderArgs
+// - ClientLoaderArgs
+// - ComponentArgs
+
+export function loader({ params }: LoaderArgs) {}
+
+export function clientLoader({ params, serverLoader }: ClientLoaderArgs) {}
+
+export default function Component({
+  params,
+  loaderData,
+  actionData,
+}: ComponentArgs) {
+  // ...
+}
+```
+
+We'll keep those hooks around for backwards compatibility, but eventually the aim is to deprecate and remove them.
+We can design new, typesafe alternatives for any edge cases.
+
+### Typegen
+
+While React Router will default to programmatic routing, it can easily be configured for file-based routing.
+That means that sometimes route URLs will only be represented as file paths.
+Unfortunately, TypeScript cannot use the filesystem as part of its type inference nor type checking.
+The only tenable way to infer types based on file paths is through code generation.
+
+To that end, React Router will generate types for each route module into a special, gitignored `.react-router` directory.
+For example:
+
+```txt
+- .react-router/
+  - types/
+    - app/
+      - routes/
+        - +types.product-details.ts
+- app/
+  - routes/
+    - product-details.tsx
+```
+
+The path within `.react-router/types` purposefully mirrors the path to the correspond route module.
+By setting things up like this, we can use `tsconfig.json`'s [rootDirs](https://www.typescriptlang.org/tsconfig/#rootDirs) option to let you conveniently import from the typegen file as if it was a sibling:
+
+```ts
+// app/routes/product-details.tsx
+import * as Types from "./+types.product-details";
+```
+
+TypeScript will even give you import autocompletion for the typegen file and the `+` prefix helps to distinguish it as a special file.
+Big thanks to Svelte Kit for showing us that `rootDirs` trick!
+
+### TypeScript plugin
+
+Typegen gets a bad rap because its a hassle when typegen'd files are out-of-date.
+Most typegen solutions require you to then rerun a script to update the typegen'd files.
```

Comment:
Some suggestions on rewording. This is going to be the main criticism I believe, so I think a slightly less colloquial. You definitely don't have to go with this, I just think "bad rap" could potentially come across as dismissive, whereas "receive criticism" shows that we accept the negative feedback as legitimate.


```suggestion
Typegen solutions often receive criticism due to typegen'd files becoming out of sync during development.
This happens because many typegen solutions require you to then rerun a script to update the typegen'd files.
```